### PR TITLE
Intake state machine updates

### PR DESCRIPTION
**Pull request by OliverW10**

### CODE REVIEW

**Comment by james-ward** [comment]
I know this is a draft, but it actually looks pretty good to me except for two things:
* The `clamping` state should be removed - all it will do is jam the cube
* Some of the state machines have been commented out in `testPeriodic()` - this is incorrect

File: controllers/acquire_cube.py [hunk 1]
```diff
@@ -57,6 +64,12 @@ def grabbing(self) -> None:
         self.gripper.close(GamePiece.CUBE)
         if self.gripper.get_full_closed():
             self.status_lights.cube_onboard()
+            self.next_state("raising")
+
+    @state(must_finish=True)
+    def raising(self):
+        self.arm.go_to_setpoint(Setpoints.HANDOFF, do_retract=False)
+        if self.arm.at_goal():
```
Comment (LucienMorey) [inline controllers/acquire_cube.py:72]: I don't really understand what adding this will do. It will already be at the handoff setpoint. Wont issuing this again with ```do_retract=False``` mean that nothing happens or have I assumed the incorrect functionality in ```arm.go_to_setpoint```. 

TLDR: I think this whole hunk can probably go because it does nothing

File: controllers/acquire_cube.py [hunk 2]
```diff
@@ -48,6 +48,13 @@ def waiting_for_cube(self) -> None:
             self.intake.deploy_without_running()
             self.next_state("grabbing")
```
Comment (LucienMorey) [inline controllers/acquire_cube.py:49]: @james-ward why was this changed back to skip the ```clamping``` state? Did we change our mind after I left on Monday? If we don't want this anymore then the whole clamping state can be removed

**Review by LucienMorey — CHANGES_REQUESTED** [review]

File: controllers/acquire_cube.py (hunk 2 above)
Comment (james-ward) [reply controllers/acquire_cube.py:49]: Clamping cannot work unless the intake is moved forwards. I left the unused state there in case we want it in the future.

File: controllers/acquire_cube.py (hunk 1 above)
Comment (james-ward) [reply controllers/acquire_cube.py:72]: If you let the recover controller move the arm, it will try to retract first which pulls the cube out of the end effector when it gets stuck on the new backboard of the well.

File: controllers/acquire_cube.py (hunk 2 above)
Comment (LucienMorey) [reply controllers/acquire_cube.py:49]: Cannot work? What do you mean by that? does it put holes in the cube or something?

Comment (james-ward) [reply controllers/acquire_cube.py:49]: You can't pull the cube up with the intake retracted.

File: controllers/acquire_cube.py (hunk 1 above)
Comment (LucienMorey) [reply controllers/acquire_cube.py:72]: But isn't that going to happen anyway because it's already in the handoff position by the time it begins this state and its told to go there again? It will just remain static and won't rotate will it? if you want to get past the intake without hitting the backboard then it sounds to me like you want the following:
```suggestion
            self.next_state("raising")

    @state(must_finish=True)
    def raising(self):
        self.arm.go_to_setpoint(Setpoints.STOW, do_retract=False)
        if self.arm.at_goal():
```

This way it will rotate outside of the collision zone and then once the recover state machine engages, it will effectively just retract

**Review by LucienMorey — CHANGES_REQUESTED** [review]

File: controllers/acquire_cube.py (hunk 1 above)
Comment (LucienMorey) [reply controllers/acquire_cube.py:72]: But isn't that going to happen anyway because it's already in the handoff position by the time it begins this state and its told to go there again? It will just remain static and won't rotate will it? if you want to get past the intake without hitting the backboard then it sounds to me like you want the following:
```suggestion
            self.next_state("raising")

    @state(must_finish=True)
    def raising(self):
        self.arm.go_to_setpoint(Setpoints.STOW, do_retract=False)
        if self.arm.at_goal():
```

This way it will rotate outside of the collision zone and then once the recover state machine engages, it will effectively just retract

Comment (OliverW10) [reply controllers/acquire_cube.py:72]: Yes, I think that was intended to go to stow

**Review by LucienMorey — APPROVED** [review]
LGTM after a few more tests to make sure its reliable

File: components/intake.py [hunk 3]
```diff
@@ -39,7 +44,24 @@ def retract(self) -> None:
...
+        # has been in same state for some time, current state is closed and wasn't only just closed
+        return (
+            (time.monotonic() - self.change_time) >= Intake.CLOSE_TIME_THRESHOLD
+        ) and self._last_deployed is self.deployed is False
+
+    @feedback
+    def is_fully_deployed(self) -> bool:
+        return (
+            (time.monotonic() - self.change_time) >= Intake.OPEN_TIME_THRESHOLD
+        ) and self.deployed is self._last_deployed is True
```
Comment (auscompgeek) [inline components/intake.py:58]: nit:
```suggestion
        ) and self.deployed and self._last_deployed
```

File: components/intake.py [hunk 4]
```diff
@@ -39,7 +44,24 @@ def retract(self) -> None:
         self.running = False
         self.deployed = False
 
+    @feedback
+    def is_fully_retracted(self) -> bool:
+        # has been in same state for some time, current state is closed and wasn't only just closed
+        return (
+            (time.monotonic() - self.change_time) >= Intake.CLOSE_TIME_THRESHOLD
+        ) and self._last_deployed is self.deployed is False
```
Comment (auscompgeek) [inline components/intake.py:52]: nit:
```suggestion
        ) and not (self._last_deployed or self.deployed)
```

**Review by auscompgeek — APPROVED** [review]
LGTM!
